sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala

```diff
@@ -491,9 +491,13 @@ object TypeCoercion {
           i
         }
 
-      case i @ In(a, b) if b.exists(_.dataType != a.dataType) =>
-        findWiderCommonType(i.children.map(_.dataType)) match {
-          case Some(finalDataType) => i.withNewChildren(i.children.map(Cast(_, finalDataType)))
+      case i @ In(value, list) if list.exists(_.dataType != value.dataType) =>
```

Comment:
@wangyum, are we able to add a legacy configuration? with it, I think it's good to go.

Reply:
Added a new config: `spark.sql.legacy.inPredicateFollowBinaryComparisonTypeCoercion`.